[charts] Support animation (#9926)
mui/mui-x: docs/data/charts/bars/bars.md

If you're using [composition](/x/react-charts/#multiple-charts), you should set the property `layout: 'horizontal'` to each bar series object.

{{"demo": "HorizontalBars.js"}}

## Animation

> To animate the creation and update of your chart use the `animate` prop.

Reviewer: ```suggestion
To animate the creation and update of your chart you can use the `animate` prop.
```

> To animate the creation and update of your chart use the `animate` prop.

Reviewer 2: Have you considered making the `animation` opt-out instead of `opt-in`? 🤔 
It looks like most libraries do have animation toggled on by default. 🤷

Author: The reason is that I'm not fully sure about the solution. I prefer to have it `opt-in` such that users willing to get animation can easily add it, and report them if there is some. Other who don't care don't get bothered with them.

Until the end of v7 beta it will let more than 3 months to battle test those animation and them move them `opt-out`

Reviewer 2: > The reason is that I'm not fully sure about the solution.

Then my suggestion is even stronger, I'd definitely go with `opt-out` to increase the time to problem discovery. 🤔 
I think that very few users might want to use charts without animation, although, I might be mistaken. 🙈

Author: 18k people disagree with you 😁

But that's a good point 🤔

Reviewer 2: We don't know how many of those downloads end up converting into usages in production environments. 🤔 
Sure, for admin panels, it's no concern, I was referring more toward regular user-facing applications. 🙈

> When set to `true` it activate animation using `@react-spring/web`.

Reviewer: ```suggestion
When set to `true` it activates animation using `@react-spring/web`.
```

Reviewer 2: ```suggestion
When set to `true` it activates animation powered by `@react-spring/web`.
```

```jsx
// For a single component chart
<BarChart {/* ... */} animate />

// For a composed chart
<ResponsiveChartContainer>
  {/* ... */}
  <BarPlot  animate />
</ResponsiveChartContainer>

> {{"demo": "BarAnimation.js"}}
> ```

Reviewer: ````suggestion
```
{{"demo": "BarAnimation.js"}}
````

If you're using [composition](/x/react-charts/#multiple-charts), you should set the property `layout: 'horizontal'` to each bar series object.

{{"demo": "HorizontalBars.js"}}

## Animation

> To skip animation at the creation and update of your chart you can use the `skipAnimation` prop.

Reviewer 2: Have you considered naming the prop `disableAnimation`?
Or is it technically incorrect?
I'd wager that the `disable` name would be way more intuitive in this case. 🤔

Author: The `skipAnimation` comes from the [`react-spring` notation](https://www.react-spring.dev/docs/utilities/use-reduced-motion#stopping-animations).

```js
Globals.assign({
  skipAnimation: false,
})
```

Not sure about why it would be more intuitive with disable

Author: The `skipAnimation` comes from the [`react-spring` notation](https://www.react-spring.dev/docs/utilities/use-reduced-motion#stopping-animations).

```js
Globals.assign({
  skipAnimation: false,
})
```

Not sure about why it would be more intuitive with disable

Reviewer 2: Thank you for clarifying.
Yeah, in such a case, we can go with `skip` instead of `disable`.
I was thinking of the global `MUI` codebase scope, where we usually use `disable` for similar props/cases. 🤔

When set to `true` it skips animation powered by `@react-spring/web`.

> Charts containers already use the `useReduceMotion` from `@react-spring/web` to skip animation [according to user preferences](https://react-spring.dev/docs/utilities/use-reduced-motion#why-is-it-important).

Reviewer 2: Maybe also mention, that when using independent chart components, it's up to the user to call this hook? 🤔

Author: This hook is called in the ChartContainer so except if they create ther own charts container they have nothing to do. 

But to do that, they will have to know very well the codebase 😅

If you're using [composition](/x/react-charts/#multiple-charts), you should set the property `layout: 'horizontal'` to each bar series object.

{{"demo": "HorizontalBars.js"}}

> ## Animation

Author: ```suggestion
## Animation

To animate the creation and update of your chart use the `animate` prop.
When set to `true` it activate animation using `@react-spring/web`.

```jsx
// For a single component chart
<BarChart {/* ... */} animate />

// For a composed chart
<ResponsiveChartContainer>
  {/* ... */}
  <BarPlot  animate />
</ResponsiveChartContainer>
```

When you activate animation, use their [`useReduceMotion`](https://www.react-spring.dev/docs/utilities/use-reduced-motion) hook for accessibility reasons.
